views-react/src/main/java/io/micronaut/views/react/ReactViewsRenderer.java

```diff
@@ -58,40 +58,41 @@ class ReactViewsRenderer<PROPS> implements ViewsRenderer<PROPS, HttpRequest<?>>
      * @param request  The HTTP request object.
      */
     @Override
-    public @NonNull Writable render(@NonNull String viewName, @Nullable PROPS props, @Nullable HttpRequest<?> request) {
-        return writer -> {
-            try {
-                beanPool.useContext(handle -> {
-                    render(viewName, props, writer, handle.get(), request);
-                    return null;
-                });
-            } catch (BeanInstantiationException e) {
-                throw e;
-            } catch (Exception e) {
-                // If we don't wrap and rethrow, the exception is swallowed and the request hangs.
-                throw new MessageBodyException("Could not render component " + viewName, e);
-            }
-        };
+    public @NonNull Mono<String> render(@NonNull String viewName, @Nullable PROPS props, @Nullable HttpRequest<?> request) {
+        StringBuilder sb = new StringBuilder();
+        RenderCallback cb;
+        try {
+            cb = beanPool.useContext(handle -> render(viewName, props, sb, handle.get(), request));
```

Comment:
I'm not sure this line is correct. Won't this release the context back to the pool when render returns, but now render is async the context should be kept open until the moment the callback is finished. At the moment no test will reveal the problem because it will only happen if you actually use `await` inside a render script.

I think rather than `useContext` the handle should be obtained using `BeanPool.checkOut` and then the returned handle can be closed inside the implementation of `callback.complete()`.